### PR TITLE
Minor Annotations Improvements 

### PR DESCRIPTION
**Pull request by OBarronCS**
This PR makes small miscellaneous changes and improvements to code relevant to annotations code, fixes some inconsistencies, updates some stale comments, and makes minor changes recommended in previous PRs. The result makes stepping through Arm, AArch64, and RISC-V much more smooth and produces more accurate disassembly in many situations.

For explanation and results on changes, see comments.

### CODE REVIEW

File: pwndbg/emu/emulator.py [hunk 1]
```diff
@@ -171,6 +171,13 @@ def debug(debug_type, fmt, args=()) -> None:
     U.UC_ARCH_RISCV: [C.riscv_const.RISCV_INS_ECALL],
 }
 
+# We stop emulation when hitting these instructions, since they depend on co-processors or other information
+# unavailable to the emulator
+BANNED_INSTRUCTIONS = {
+    "mips": {C.mips.MIPS_INS_RDHWR},
+    "arm": {C.arm.ARM_INS_MRC, C.arm.ARM_INS_MRRC, C.arm.ARM_INS_MRC2, C.arm.ARM_INS_MRRC2},
+}
```
Comment (OBarronCS) [inline pwndbg/emu/emulator.py:179]: Unicorn will normally emulate through these instructions, but get the incorrect results. They relate to receiving some value from a "coprocessor" and placing it into a register.

Fun fact: In MIPS, the `rdhwr` is seen often querying register number 29 which returns the thread local storage base: https://s3-eu-west-1.amazonaws.com/downloads-mips/documents/MIPS_Architecture_MIPS64_InstructionSet_%20AFP_P_MD00087_06.05.pdf#page=390

Likewise, MRC on Arm is used to read the TLS base.

The emulator just reads zeros into the registers, causing the subsequent instruction to be inaccurate.

File: pwndbg/emu/emulator.py [hunk 2]
```diff
@@ -626,7 +633,8 @@ def hook_intr(self, uc, intno, user_data) -> None:
         """
         We never want to emulate through an interrupt.  Just stop.
         """
-        debug(DEBUG_INTERRUPT, "Got an interrupt")
+        debug(DEBUG_INTERRUPT, "Got an interrupt - %d", intno)
+        self.valid = False
```
Comment (OBarronCS) [inline pwndbg/emu/emulator.py:637]: Previously, an interrupt during a single_step would not cause the step to fail. We set `valid` to False to force it to fail.

File: pwndbg/emu/emulator.py [hunk 3]
```diff
@@ -810,25 +814,28 @@ def single_step(self, pc=None, check_instruction_valid=True) -> Tuple[int, int]:
...
+            DEBUG_EXECUTING,
+            "# Emulator attempting to single-step at %#x: %s %s",
+            (pc, insn.mnemonic, insn.op_str),
+        )
 
         try:
             self.single_step_hook_hit_count = 0
             self.emulate_with_hook(self.single_step_hook_code, count=1)
+            if not self.valid:
+                return InstructionExecutedResult(None, None)
```
Comment (OBarronCS) [inline pwndbg/emu/emulator.py:838]: Checks for interrupt during the instruction, and returns failure in this case

File: pwndbg/gdblib/disasm/__init__.py [hunk 4]
```diff
@@ -386,6 +386,8 @@ def near(
         while insn is not None and len(insns) < instructions:
             if DEBUG_ENHANCEMENT:
                 print(f"Got instruction from cache, addr={cached:#x}")
+            if insn.jump_like and insn.split == SplitType.NO_SPLIT:
+                insn.split = SplitType.BRANCH_NOT_TAKEN
```
Comment (OBarronCS) [inline pwndbg/gdblib/disasm/__init__.py:390]: Without this, branches in the disasm view that are before the current instruction pointer, but have all information cleared from the cache (possible in multiple scenarios) would have no split in the disasm view. This adds the split.

File: pwndbg/gdblib/disasm/__init__.py [hunk 5]
```diff
@@ -423,7 +425,7 @@ def near(
 
         # Handle visual splits in the disasm view
         # The second check here handles instructions like x86 `REP` that repeat the instruction
-        if insn.jump_like or insn.next == insn.address:
+        if insn.has_jump_target or insn.next == insn.address:
```
Comment (OBarronCS) [inline pwndbg/gdblib/disasm/__init__.py:428]: Results in instructions in Arm such as 
```
pop pc
ldr pc, [some_address]
```
To have split in disasm view

File: pwndbg/gdblib/disasm/aarch64.py [hunk 6]
```diff
@@ -187,9 +187,13 @@ def _condition(
...
-            flags = super()._read_register_name(instruction, "cpsr", emu)
-            if flags is not None:
-                return resolve_condition(instruction.cs_insn.cc, flags)
+            # The B instruction can be made conditional by the condition codes
+            if instruction.cs_insn.cc in (ARM64_CC_INVALID, ARM64_CC_AL):
+                instruction.declare_conditional = False
+            else:
+                flags = super()._read_register_name(instruction, "cpsr", emu)
+                if flags is not None:
+                    return resolve_condition(instruction.cs_insn.cc, flags)
```
Comment (OBarronCS) [inline pwndbg/gdblib/disasm/aarch64.py:196]: AArch64 encodes many conditional branch instructions under the same instruction opcode, `B`. This change removes the checkmark from a plain `B` (it is unconditional).

File: pwndbg/gdblib/disasm/arch.py [hunk 7]
```diff
@@ -667,19 +670,15 @@ def _enhance_next(
...
-    def _resolve_target(self, instruction: PwndbgInstruction, emu: Emulator | None, call=False):
+    def _resolve_target(self, instruction: PwndbgInstruction, emu: Emulator | None):
         """
         Architecture-specific hook point for _enhance_next.
 
-        Returns the value of the instruction pointer assuming this instruction executes (and any conditional jumps are taken)
-
-        "call" specifies if we allow this to resolve call instruction targets
+        Returns the program counter target of this instruction.
+        Even in the case of conditional jumps, the potential target should be resolved.
```
Comment (OBarronCS) [inline pwndbg/gdblib/disasm/arch.py:678]: Old comment was incorrect. The determination of whether the _resolve_target can handle `call` instructions is now resolved by the caller of this function (and all subclasses now reflect this), which makes it much simpler and correct.

File: pwndbg/gdblib/disasm/arm.py [hunk 8]
```diff
@@ -91,6 +91,8 @@ def _set_annotation_string(self, instruction: PwndbgInstruction, emu: Emulator)
     @override
     def _condition(self, instruction: PwndbgInstruction, emu: Emulator) -> InstructionCondition:
         if instruction.cs_insn.cc == ARM_CC_AL:
+            if instruction.id in (ARM_INS_B, ARM_INS_BL, ARM_INS_BLX, ARM_INS_BX, ARM_INS_BXJ):
+                instruction.declare_conditional = False
```
Comment (OBarronCS) [inline pwndbg/gdblib/disasm/arm.py:95]: Like in AArch64, these Arm instructions can be conditional based on the condition code. This change reflects this fact in the disasm view (checkmarks are rationed out correctly, and jumps are unrolled correctly now)

File: pwndbg/gdblib/disasm/instruction.py [hunk 9]
```diff
@@ -53,16 +52,18 @@
...
     CS_ARCH_ARM64: {ARM64_INS_BL, ARM64_INS_BLR, ARM64_INS_BR},
-    CS_ARCH_RISCV: {RISCV_INS_JAL, RISCV_INS_JALR},
+    CS_ARCH_RISCV: {
+        RISCV_INS_JAL,
+        RISCV_INS_JALR,
+        RISCV_INS_C_JAL,
+        RISCV_INS_C_JALR,
+        RISCV_INS_C_J,
+        RISCV_INS_C_JR,
+    },
```
Comment (OBarronCS) [inline pwndbg/gdblib/disasm/instruction.py:66]: Add missing unconditional jump types

File: pwndbg/gdblib/disasm/instruction.py [hunk 10]
```diff
@@ -232,6 +233,21 @@ def __init__(self, cs_insn: CsInsn | None) -> None:
         FALSE if the instruction has a conditional action, and we know it is not taken.
         """
 
+        self.declare_conditional: bool | None = None
+        """
```
Comment (OBarronCS) [inline pwndbg/gdblib/disasm/instruction.py:237]: New field necessary for handling Arm/Arch branch instruction conditionality as they relate to how we unroll jumps in the disasm view.

File: pwndbg/gdblib/disasm/riscv.py [hunk 11]
```diff
@@ -137,57 +140,46 @@ def _is_condition_taken(
...
-
         # Determine if the conditional jump is taken
         if RISCV_GRP_BRANCH_RELATIVE in instruction.groups:
             return self._is_condition_taken(instruction, emu)
 
         return InstructionCondition.UNDETERMINED
 
     @override
-    def _resolve_target(self, instruction: PwndbgInstruction, emu: Emulator | None, call=False):
+    def _resolve_target(self, instruction: PwndbgInstruction, emu: Emulator | None):
```
Comment (OBarronCS) [inline pwndbg/gdblib/disasm/riscv.py:157]: All RISC-V jump instructions now resolve targets.

**Comment by disconnect3d** [comment]
LGTM. Any chnace to add some tests for this?

**Pull request merged**